### DISCUSSION TITLE
.umi 文件夹中包含了一些导出的方法，这些方法是如何做到映射到 umi 的包中去的

### DISCUSSION THREAD

**Discussion by qiugu** [umijs/umi#11821] [Q&A]
umi 中可以导入 import { useModel } from 'umi'，这里 useModel 实际是 .umi 中的 model 插件下的 index.ts 文件中导出的，然后 exports 文件中引入了，同时 umi 的 类型文件 index.d.ts 中加入了 export * from '@@/exports' 指向了这个临时文件 exports.ts，但是实际 umi 中不是应该没有这个导出吗，还是说通过 cjs 的运行时能力，运行时加载了 exports.ts 中的内容？看不懂啊，来个大佬解惑一下，拜谢

**Comment by fz6m** [comment]
运行时 webpack 里 alias 把 `umi` 指向了 `src/.umi/exports.ts` 。

**Comment by qiugu** [comment]
是bundler-webpack这个包里面吗，我看到源码里面，直接拿的配置文件里面的alias，有另外单独定义的吗

**Reply by fz6m** [reply]
有关具体实现，可自行查询源码。

**Reply by qiugu** [reply]
最后一个问题，这里umi导出了很多 Node 端的代码，然后还包括了 exports 导出的浏览器端的代码，为啥我引用umi的时候，就会报错，不引用就没有问题

**Reply by fz6m** [reply]
在 webpack 的项目中，绝大概率不会有 node 端的代码，在 umi 项目中也是没有 node 端代码的，都是纯浏览器运行的代码。
